### ISSUE TITLE
Some concrete files in the test suite fail (printing issues?)

### ISSUE THREAD

**Issue by GenevaS** [hakaru-dev/hakaru#89]
Some tests are failing even though they are equivalent. Here is an example:

````bash
### Failure in: 6:RoundTrip:3:5:dup
    haskell\Tests\TestTools.hs:90
    expected:
    x1 <~ normal(nat2real(0), nat2prob(1))
    x0 <~ normal(nat2real(0), nat2prob(1))
    return (x1, x0)
    but got:
    x15 <~ normal(0, 1)
    x03 <~ normal(0, 1)
    return (x15, x03)
````

Possibly the result of incorrect pretty printing?

**Comment by JacquesCarette** [comment]
Is the 'got' program now correct Hakaru?  @zaxtax ?  If so, then that is indeed better and should be accepted as the new expected.

**Comment by zaxtax** [comment]
The got one looks fine.

**Comment by yuriy0** [comment]
This seems like an error with alpha equality. If the latter program contains `LProb 1` while the former contains `CoerceTo HProb (LNat 1)` then those should be considered equal.

**Comment by JacquesCarette** [comment]
Yes, but I don't want to reproduce the coercion logic in alpha equality.  So, until this becomes a real pain, I think we should just accept the 'got' program as the new, nicer, 'expected'. 

Furthermore, this does represent a real improvement, so it is actually *good* that we were forced to look at it.  If things had moved in the opposite direction, it should have been a regression.

**Comment by zaxtax** [comment]
I agree about that suggestion.

**Comment by yuriy0** [comment]
In reality, the issue is that the pretty printer is throwing away information. A change to alpha equality may fix some of these, but not all. This failure (and about 30 others) begin happening as a result of having moved the RoundTrip tests to concrete syntax. 

As a simpler example, `RoundTrip.t82'` is
```
syn (Lam_ :$ ((bind (Variable "" (Nat 0) SReal) (syn (Literal_ (LProb (1 % 1))))) :* End))
```
whereas reading this same program from concrete syntax (i.e. `fmap parseAndInfer (IO.readFile "tests/RoundTrip/t82.expected.hk")`) is 
```
syn (Lam_ :$ ((bind (Variable "x0" (Nat 0) SReal) (syn (Literal_ (LNat 1)))) :* End)))
```
There is no way for `alphaEq` to ever decide these are the same program, because they have different types. Even worse, the latter program is the result of pretty printing the former program. In other words, the pretty printer has discarded type information.

I'm pretty sure that #90 is the same issue, but manifested in a slightly different way. 

> I think we should just accept the 'got' program as the new, nicer, 'expected'.

There is no new program. This program is just the old one with an additional roundtrip through `parseAndInfer . pretty`. It turns out this operation is not idempotent but I feel confident that it should be.

**Comment by zaxtax** [comment]
So the issue is that parsing adds coercions that pretty than preserves. For
example,

$ echo "uniform(0,1)" | pretty
uniform(nat2real(0), nat2real(1))

So maybe Pretty/Concrete should throw away coercions, or we should start
with coercions.

**Comment by ccshan** [comment]
First of all, a Hakaru program is an abstract syntax tree that may contain coercions. For example, the program `nat2prob(5) + nat2prob(30)` is different from the program `nat2prob(5 + 30)` (even though both have the type `prob` only), and running the two programs through the `hakaru` command produces slightly different outputs.

Sometimes parsing a concrete syntax file that contains no explicit coercions nevertheless produces a Hakaru program (i.e., an abstract syntax tree) that contains coercions. Rob's program `uniform(0,1)` is a fine example. That program (i.e., that abstract syntax tree) is the result of parsing an infinite number of different concrete syntax files:
* `uniform(0,1)`
* `uniform(nat2real(0), nat2real(1))`
* `uniform(0, nat2real(1))`
* `uniform(nat2real(0), 1)`
* `uniform(0,             00001)`
* etc.
Let's call this program P.

If we take a Hakaru program (i.e., an abstract syntax tree), pretty-print it, then parse the resulting text, we should get back the same Hakaru program (i.e., an alpha-equivalent abstract syntax tree). In particular, we should get back a Hakaru program with exactly the same coercions in the same places. For example, the result of pretty-printing P can be any of the infinite number of texts listed above, because they all parse back to P. In contrast, pretty-printing `nat2prob(5) + nat2prob(30)` and pretty-printing `nat2prob(5 + 30)` must give different results, because the results must parse back to different programs (i.e., different abstract syntax trees, with different coercions in different places). The easiest way to ensure that coercions don't interfere with this roundtripping is for pretty-printing to make all coercions explicit and for parsing to add no implicit coercions as long as the explicit coercions are enough to make the program type-check. That's what we currently do, right?

The problem Yuriy demonstrated is that the programs `syn (Literal_ (LProb (1 % 1)))` and `syn (Literal_ (LNat 1))` break this roundtripping: pretty-printing them gives the same result, so it's impossible for the result to parse back to different programs. To fix this problem, we should introduce concrete notation for `prob` literals distinct from `nat` literals (and similarly for `real` and `int` literals), and make sure that the pretty-printer and the parser agree on this notation.

**Comment by zaxtax** [comment]
If you think we need literals for prob and nat, add an issue about it.

**Comment by JacquesCarette** [comment]
Between this and #89, we have an accumulation of failing tests now stuck on parsing/pretty-printing issues.  These tests do not seem outlandish to me -- so this seems like a fairly serious issue. @yuriy0 is working on fixing issues on the Maple side (also causing failures of tests we would really like to pass); could perhaps @zaxtax or @ccshan look at the Haskell-side of things?

**Comment by ccshan** [comment]
(Just to clarify, between **what** and #89? This is #89.)

**Comment by JacquesCarette** [comment]
I've lost track.  I'll get @yuriy0 to update the test failure classification (it should be in the repo) to make that clearer.

**Comment by yuriy0** [comment]
This change (1d40bdebd84fd293622e1f2ed632037213c994db) will make sure a program isn't pretty printed in a way which throws away type information (although, `1` and `1.prob` and `1.real` should all 'unify' anyways). It still isn't strong enough to make all the new failures go away. Note that #90 is unaffected by any of these recent changes - it seems based on the discussion there that that is an issue with type checking/inference, whereas this one is an issue with representation of equivalent programs.

>  update the test failure classification

This will help to identify what still needs to be done.

**Comment by yuriy0** [comment]
> To fix this problem, we should introduce concrete notation for prob literals distinct from nat literals (and similarly for real and int literals), and make sure that the pretty-printer and the parser agree on this notation.

This would help the problem; but I'm not certain it would completely fix it. 

If we did do this, I would propose the following syntax:
- `0`, `1` have type `nat`
- `+0`, `+1`, `-1` have type `int`. This would change the parsing of `-1` from `Negate (LNat 1)` to `LInt (-1)`.
- `0.0`, `1.0`, `1/1` have type `prob`. This would change the parsing of `1/1` from `Prod (LProb 1) (Recip (LProb 1))` to `LProb 1`. 
- `+0.0`, `+1.0`, `-1.0`, `+1/1`, `-1/1` have type `real`. This would again change the parsing of `-1.0`.


One example of something not fixed by such a change is `RoundTrip/mhPriorProp` (bb418973d2afdfaa14d6f28489c5a4a4f17ce87e), which simplifies to the expected output, but for one difference:

```
expected:
... * int2real(-1) * prob2real(recip(nat2prob(2)))
got:
... * -1/2
```

If we had a distinct syntax for each literal type, then `expected` would contain `... * -1/1 * recip(+2/1)`; `got` would still contain `... * -1/2`. 

Another thing to consider: this issue arose as the result of moving tests from Haskell syntax to concrete syntax. All the issues seem to arise from the interpretation of literals. While the programs in question were written in Haskell syntax, literal polymorphism was taken care of by the Haskell type system, using the same mechanism that Haskell uses - the literal `0` is interpreted as `fromIntegral 0 :: Num a => a`. The Hakaru typechecker uses a different implementation of polymorphism for literals - the literal `0` is alternatively interpreted as a subset of `{nat, int, prob, real}` depending on the context in which it appears; and upon completion of typechecking, a single type of the remaining possible types is chosen for the literal. If we wanted to ensure that a Hakaru program could just as easily be written in concrete syntax as in Haskell syntax, we could ensure that these typechecking algorithms agree - if not in implementation, at least in results. 

Yet another thing to consider: it may be the case that `alphaEq` should be comparing terms for semantic equality, and not for type-directed syntactic equality. In the case of a syntactic equality, `CoerceTo c0 x0` is only equal to `CoerceTo c1 x1` if `x0` and `x1` are equal; the coercions are ignored because there should be a single inhabitant of a coercion at a particular type. If the equality was instead semantic, it could take two terms of differing types, say `LInt 1` and `LReal 1` and prove they represent equal values (that is, these programs are identical when sampled from) even if their types don't agree. Such an equality, being heterogeneous, would be free to ignore coercions altogether - that is, `CoerceTo c0 x0` and `x1` are equal if `x0` and `x1` are equal. The old 'type-directed syntactic' equality could be recovered simply by instantiating the heterogeneous equality to takes two terms of equal type.

**Comment by ccshan** [comment]
I like the literal syntax you propose, except I'm worried that `/` also means division, so how do we express the division of 1 by 2 in concrete syntax? I hope it's not just by putting spaces around `/` in `1 / 2`. Maybe by putting parentheses in `(1)/2` or `1/(2)`? Whatever disambiguation we use would need to be generated by the pretty printer, of course.

I also hope spaces would be allowed around not only `/` but also `+` and `-` in this literal syntax.

Due to precedence, I'm not sure that `... * -1/2` should be allowed, even if `-1/2` is parsed as a literal.

Regarding your "another thing to consider": It's probably too big a change for now, because
1. it would be a significant change in interface -- for example, the user would have to insert an explicit coercion in the following addition of a `nat` to a `real`:
```
i <~ poisson(42)
x <~ normal(0,1)
return (i+x)
```
2. it would be a significant change in implementation -- in particular, type-class resolution calls for type variables and unification.

Regarding your "yet another thing to consider": You seem to be talking about making our testing use a coarser notion of equality. If I understand you correctly, the reason to do it is, if a test currently produces A (and passes) and the result changes to B in the future, and we don't care about the difference between A and B, then a coarser notion of equality that equates A and B would save us from the bother of updating the expected test result. I get the impression that the current test failures are due to the move to concrete syntax, so I haven't seen examples of A and B, and I would wait for them.

**Comment by yuriy0** [comment]
> I'm worried that / also means division, so how do we express the division of 1 by 2 in concrete syntax?

In my mind, that term would become inexpressible with such a change. Is there a need to be able to express it? That is, is there any situation where the literal `1/2` doesn't serve the same purpose as division of 1 by 2?

> Due to precedence, I'm not sure that ... * -1/2 should be allowed, even if -1/2 is parsed as a literal.

Due to precedence `.. * -1/2` is parsed as `Prod (.., -1, Recip(2))`. But that term (`... * -1/2` in the `got` of `mhPriorProp`) comes from parsing Maple, not parsing concrete syntax. I believe that the Maple parser is what produces `Prod(.., -1/2)` - the Maple parser receives the result of `ToInert`, which picks its own association of operations. That association may not agree with what the `Concrete` parser would pick if it parsed the pretty printed program. Maybe that's a part of the problem!

If `-1/2` were a literal according to the scheme I mentioned, I think that precedence would no longer apply once parsing `-1/2` because it would be a single 'atom'. However, `.. * -1/2` could still be parsed as `Prod(.., -1, Recip(2))`, if the parser entered a different context upon parsing a `*` (or another operator). I'm not sure this would be necessary - do we lose anything by treating `-,/,+` differently when they are applied to literals? Treating them differently was part of my goal in proposing the new syntax; by doing so, we could ensure a canonical form of literals right in the parser. 

> It's probably too big a change for now...

I agree with both reasons. I did not even consider the first, and can't see off-hand how it could be solved (that is, to not place additional burden on the user); as for the second, it was evident, but I had hoped there may be a way to implement something of the sort without having to implement type variables in the internal syntax, and without having to implement a unification algorithm (i.e. an implementation which emulates type class unification "not in implementation, [but] at least in results."). 

> If I understand you correctly, the reason to do it is, if a test currently produces A (and passes) and the result changes to B in the future, and we don't care about the difference between A and B, then a coarser notion of equality that equates A and B would save us from the bother of updating the expected test result.

This isn't my reason, although it would be a nice side effect! To clarify my goal there: currently, some equivalent programs (and not just equivalent in the sense that every simplified program is equivalent to the original; equivalent up to a very trivial amount of differences) are emitted by different parts of Hakaru (e.g. the Maple parser and concrete parser) and we should not have to ensure every single part of Hakaru emits the 'same' program in identical internal syntax. Such equivalence classes of programs are very apparently the same program to the human eye, and they almost certainly are equally as efficient as samplers (if there is a discrepancy, I think it would be absolutely insignificant; multiplying by -1 then dividing by 2 is basically the same as multiplying by -1/2). 

The only time a machine needs to determine two programs are the same program is inside `alphaEq`. So it is `alphaEq` that should ensure the equivalence of morally equivalent programs - all of the parsers, pretty printers, and any other part of Hakaru, should not have to work in perfect harmony just to "prove" a trivial equivalence. 

> save us from the bother of updating the expected test result

To put it another way, it is not simply a bother to update the expected result; for some results, it is impossible to write that result in concrete syntax. We could expand the parser to allow the user to write *all* programs in concrete syntax (that is, every possible equivalent variation of a particular program should have a syntax which parses to that exact variation); or we could change `alphaEq` to judge all such equivalent programs equal. The latter seems like a better choice to me, because there are many parsers, but one `alphaEq`.

**Comment by ccshan** [comment]
The coarser our equivalence relation on programs, the more we have to worry that a transformation might behave differently on two equivalent programs.  For example, I'm not 100% sure offhand whether disintegration treats a fractional literal and a division of two literals identically.  We already have this worry given that we compare programs modulo renaming variables and permuting n-ary operands, so the worry is not new, but it can get worse.

I still stand by what I wrote above:

> If we take a Hakaru program (i.e., an abstract syntax tree), pretty-print it, then parse the resulting text, we should get back the same Hakaru program (i.e., an alpha-equivalent abstract syntax tree).  In particular, we should get back a Hakaru program with exactly the same coercions in the same places.

This requirement entails that no term be inexpressible, because the job of the pretty-printer is to find a way to express every term.  Of course we could shrink the AST data type to rule out terms we don't feel like expressing, but I don't see how to do that in this case.

I'm ok with literals that look like compound expressions (such as `-1/2`), as long as the syntax of these literals is consistent with the usual precedence and associativity of the operators involved (`+`, `-`, `/`) and remains insensitive to whitespace and comments.  So `1 / 2 / 3 / 4` should be parsed as the literal `1/2` divided by `3` then divided by `4`, not as the literal `1/2` divided by the literal `3/4`, which would yield a different number.  And `... * -1 / 2` should not be parsed as multiplying `...` by the literal `-1/2`, because `*` and `/` have the same precedence are left-associative.  Actually, I'd rather `... * -1 / 2` not parse at all, because I don't like unary `-` in the non-leftmost operand of `*` -- it's not standard math and it's not allowed by Haskell (see `lexp⁶ → - exp⁷` in https://www.haskell.org/onlinereport/exps.html).

Parentheses, such as in `-(1)/2`, can be used (in particular, by the pretty-printer) to make non-literals expressible.

**Comment by yuriy0** [comment]
> the more we have to worry that a transformation might behave differently on two equivalent programs

Indeed; it seems that changing alphaEq would be at least somewhat burdensome, and wouldn't obviate the other necessary changes to the parser/pretty printer. I'm now convinced this isn't the way to go.


To summarize, I see this issue as the following distinct issues:

- The concrete parser and pretty-printer should implement the new syntax for literals. This should include a way of specifying e.g. division of two literals which would normally be parsed as a single literal. I agree that parentheses would be a good way to do that. I also agree that non-indentation whitespace should not have an effect on parsing, in general, but in particular `1/2` and `1  /  2` should denote exactly the same term.

- The pretty printer and parser should handle precedence correctly. I would also rather not have `... * -1/2` parse; but it currently does. This includes making sure the 'overloading' of `+`, `-`, `/` in the new literal syntax agree with the precedences of existing operators. This issue may also include other precedence related problems (https://github.com/hakaru-dev/hakaru/commit/3f2f7e47e66f22d54fbd522f2717431e4bede4bc#commitcomment-22789801); or that particular one might also be a separate issue. 

- The Maple parser should agree with the concrete parser on how a program is parsed. In particular, if the simplifier does not change a particular program, then `hk-maple prog.hk` should produce a program alpha-equivalent to `prog.hk`. 

- The change I implemented (e25c1eb3cf6a401b550bce60fa07c67d3752f61b) to make sure literals were not pretty printed with the incorrect type should be reverted. If each literal type has its own syntax, that change is no longer necessary, and it creates a lot of syntactic noise in the output of `pretty`, and it would (currently does) defeat the goal of having every program be expressible in concrete syntax.

**Comment by ccshan** [comment]
Maybe this is done as of e214cbe9d34fa7a43eedd104b9a5e1b4f33b191c?

**Comment by yuriy0** [comment]
As of 5e6067ff442c351769ebfacf679018c122c69f3b it is really done. Thanks for the fixes!
